components/style/animation.rs

```diff
@@ -531,6 +531,63 @@ impl ElementAnimationState {
         // Otherwise just add the new running animation.
         self.add_new_animation(new_animation);
     }
+
+    /// Update our animations given a new style, canceling or starting new animations
+    /// when appropriate.
+    pub fn update_animations_for_new_style<E>(
+        &mut self,
+        element: E,
+        context: &SharedStyleContext,
+        new_style: &Arc<ComputedValues>,
+    ) where
+        E: TElement,
+    {
+        // Cancel any animations that no longer exist in the style.
+        // TODO(mrobinson): We should make this more efficient perhaps by using
+        // a linked-list or by using something like `partition`.
+        if !self.running_animations.is_empty() {
+            let animation_count = self.running_animations.len();
+            let mut previously_running_animations = std::mem::replace(
+                &mut self.running_animations,
+                Vec::with_capacity(animation_count),
+            );
+            for animation in previously_running_animations.drain(..) {
+                if animation.is_animation_cancelled_in_new_style(new_style) {
+                    self.cancelled_animations.push(animation);
+                } else {
+                    self.running_animations.push(animation);
+                }
+            }
+        }
+
+        maybe_start_animations(element, &context, &new_style, self);
+    }
+
+    /// Update our transitions given a new style, canceling or starting new animations
+    /// when appropriate.
+    pub fn update_transitions_for_new_style(
+        &mut self,
+        context: &SharedStyleContext,
+        opaque_node: OpaqueNode,
+        before_change_style: &Option<Arc<ComputedValues>>,
```

Comment:
Maybe make this `Option<&Arc<>>`? `&Option` feels a bit weird but...

Reply:
That makes sense. The latest version changes this to use `Option<&Arc<>>`.